### DISCUSSION TITLE
How can I set value to atomFamily dynamically

### DISCUSSION THREAD

**Discussion by LvChengbin** [pmndrs/jotai#2437] [Q&A]
I push data from server to client via WebSocket, and the data structure is similar to this:

```json
{
    "messages" : {
        "id1" :  {},
        "id2" :  {},
        "id3" :  {},
        "id4" :  {},
        "id5" :  {},
        ...
        "idn" :  {},
    }
}
```
I want to save the data of each "id" separately, and use them in different components.

With the `atomFamily` function, I manage the data like this:

```ts
const messagesFamily = atomFamily( id => atom( {} ) );
```

This's the sample code that I want to describe my requirement:

```tsx
ws.on( 'message', ( messages ) => {
    Object.keys( messages ).forEach( ( id ) => {
        // How can I save the data into different atom with it's id
        setMessages( id, messages[ id ] );
    } );   
} );
```

And another question is how can I delete an atom from the atomFamily with it's id when I don't need the data any more.

**Comment by dai-shi** [comment]
You may want to use the store api.

atomFamily has `remove` method, but if it were required, I would consider using a single atom or atoms-in-atom pattern.

**Reply by LvChengbin** [reply]
Thank you for your answer and would you like to show me more details of using the atoms-in-atom pattern?

**Reply by dai-shi** [reply]
https://github.com/pmndrs/jotai/blob/c974d15fa651c540b74a2e569d6b33e40b3a8bf5/docs/guides/atoms-in-atom.mdx#L2 shows some basic ideas. Hope it helps.

Basically, the structure will be like:
```js
const messagesAtom = atom({
  "id1": atom(...),
  "id2": atom(...),
  "id3": atom(...),
  "id4": atom(...),
  ...
})
```
